lib/widgets/message_list.dart

```diff
@@ -635,3 +714,76 @@ final _kMessageTimestampStyle = TextStyle(
   fontSize: 12,
   fontWeight: FontWeight.w400,
   color: const HSLColor.fromAHSL(0.4, 0, 0, 0.2).toColor());
+
+Future<void> markNarrowAsRead(BuildContext context, ApiConnection connection, Narrow narrow) async {
+  // Compare web's `mark_all_as_read` in web/src/unread_ops.js
+  // and zulip-mobile's `markAsUnreadFromMessage` in src/action-sheets/index.js .
+  final zulipLocalizations = ZulipLocalizations.of(context);
+  // Use [AnchorCode.oldest], because [AnchorCode.firstUnread]
+  // will be the oldest non-muted unread message, which would
+  // result in muted unreads older than the first unread not
+  // being processed.
+  Anchor anchor = AnchorCode.oldest;
+  int responseCount = 0;
+  int updatedCount = 0;
+  while (true) {
+    final result = await updateMessageFlagsForNarrow(connection,
+      anchor: anchor,
+      // [AnchorCode.oldest] is an anchor ID lower than any valid
+      // message ID; and follow-up requests will have already
+      // processed the anchor ID, so we just want this to be
+      // unconditionally false.
+      includeAnchor: false,
+      // There is an upper limit of 5000 messages per batch
+      // (numBefore + numAfter <= 5000) enforced on the server.
+      // See `update_message_flags_in_narrow` in zerver/views/message_flags.py .
+      // zulip-mobile uses `numAfter` of 5000, but web uses 1000
+      // for more responsive feedback. See zulip@f0d87fcf6.
+      numBefore: 0,
+      numAfter: 1000,
+      narrow: narrow.apiEncode(),
```

Comment:
Web has the following optimization, for the mark-*all*-as-read case:
```js
        // Since there's a database index on is:unread, it's a fast
        // search query and thus worth including here as an optimization.
        narrow: JSON.stringify([{operator: "is", operand: "unread", negated: false}]),
```

That's important when there's a ton of already-read history — as there may very often be for the "all" case or the stream case. So we should do the same.

The way to express that is with another `ApiNarrowElement` subclass. Say simply `ApiNarrowIsUnread`. Can go right above `ApiNarrowMessageId`, after all the other `ApiNarrowElement` subclasses.

(Later we can later generalize that to an `ApiNarrowIs` where the operand is some appropriate enum. But that's a part of the API that's only sort of documented; so that'll involve some reverse-engineering, not a pure matter of transcribing API docs, and we can leave it for another time.)